confdb/confdb_control.go

```diff
@@ -31,134 +31,312 @@ var (
 	validAccountID = regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$")
 )
 
-// AuthenticationMethod limits what keys can be used to sign messages used to remotely
-// manage confdbs.
-type AuthenticationMethod string
+// ConfdbControl holds a lists of views delegated by the device to operators.
+type ConfdbControl struct {
+	// the key is the operator ID
+	operators map[string]*operator
+}
 
-const (
-	// Only the operator's keys can be used to sign the messages.
-	OperatorKey AuthenticationMethod = "operator-key"
-	// Messages can be signed on behalf of the operator by the store.
-	Store AuthenticationMethod = "store"
-)
+// Delegate delegates the given views with the provided auth to the operator.
+func (cc *ConfdbControl) Delegate(operatorID string, views, auth []string) error {
+	if !validAccountID.MatchString(operatorID) {
+		return fmt.Errorf("invalid operator ID: %s", operatorID)
+	}
+
+	if cc.operators == nil {
+		cc.operators = make(map[string]*operator)
+	}
+
+	op, ok := cc.operators[operatorID]
+	if !ok {
+		op = &operator{ID: operatorID}
+	}
+
+	err := op.delegate(views, auth)
+	if err != nil {
+		return err
+	}
+
+	cc.operators[operatorID] = op
+	return nil
+}
+
+// Undelegate withdraws access to the views that have been delegated with the provided auth.
+func (cc *ConfdbControl) Undelegate(operatorID string, views, auth []string) error {
+	operator, ok := cc.operators[operatorID]
+	if !ok {
+		return nil // nothing is delegated to this operator
+	}
+
+	return operator.undelegate(views, auth)
+}
+
+// IsDelegated checks if the view is delegated to the operator with the given auth.
+func (cc *ConfdbControl) IsDelegated(operatorID, view string, auth []string) (bool, error) {
+	operator, ok := cc.operators[operatorID]
+	if !ok {
+		return false, nil // nothing is delegated to this operator
+	}
+
+	return operator.isDelegated(view, auth)
+}
+
+// Groups returns the groups in a format that can be used to assemble the next revision
+// of the confdb control assertion.
+func (cc *ConfdbControl) Groups() []interface{} {
+	authMap := map[authentication]map[viewRef][]string{}
```

Comment:
probably a good idea to have a comment on this that also mentions that string at the end is operator ids